src/checkout-buttons/strategies/paypal-button-strategy.ts

```diff
@@ -0,0 +1,142 @@
+
+export default class PaypalButtonStrategy extends CheckoutButtonStrategy {
+    private _paymentMethod?: PaymentMethod;
+
+    constructor(
+        private _store: CheckoutStore,
+        private _checkoutActionCreator: CheckoutActionCreator,
+        private _paypalScriptLoader: PaypalScriptLoader,
+        private _formPoster: FormPoster
+    ) {
+        super();
+    }
+
+    initialize(options: CheckoutButtonInitializeOptions): Promise<void> {
+        if (this._isInitialized[options.containerId]) {
+            return super.initialize(options);
+        }
+
+        const paypalOptions = options.paypal;
+        const state = this._store.getState();
+        const paymentMethod = this._paymentMethod = state.paymentMethods.getPaymentMethod(options.methodId);
+
+        if (!paypalOptions) {
+            throw new InvalidArgumentError();
+        }
+
+        if (!paymentMethod || !paymentMethod.config.merchantId) {
+            throw new MissingDataError(MissingDataErrorType.MissingPaymentMethod);
+        }
+
+        const env = paymentMethod.config.testMode ? 'sandbox' : 'production';
+        const clientToken: PaypalClientToken = { [env]: paypalOptions.clientId };
+
+        return this._paypalScriptLoader.loadPaypal()
+            .then(paypal => {
+                const allowedSources = [];
+                const disallowedSources = [];
+
+                if (paypalOptions.allowCredit) {
+                    allowedSources.push(paypal.FUNDING.CREDIT);
+                } else {
+                    disallowedSources.push(paypal.FUNDING.CREDIT);
+                }
+
+                return paypal.Button.render({
+                    env,
+                    client: clientToken,
+                    commit: paypalOptions.shouldProcessPayment,
+                    funding: {
+                        allowed: allowedSources,
+                        disallowed: disallowedSources,
+                    },
+                    style: {
+                        shape: 'rect',
+                        ...pick(paypalOptions.style, 'layout', 'size', 'color', 'label', 'shape', 'tagline', 'fundingicons'),
+                    },
+                    payment: (data, actions) => this._setupPayment(paymentMethod.config.merchantId, actions, paypalOptions.onPaymentError),
+                    onAuthorize: (data, actions) => this._tokenizePayment(data, actions, paypalOptions.shouldProcessPayment, paypalOptions.onAuthorizeError),
+                }, options.containerId);
+            })
+            .then(() => super.initialize(options));
+    }
+
+    deinitialize(): Promise<void> {
+        if (!Object.keys(this._isInitialized).length) {
+            return super.deinitialize();
+        }
+
+        this._paymentMethod = undefined;
+
+        return super.deinitialize();
+    }
+
+    private _setupPayment(merchantId: string | undefined, actions?: PaypalActions, onError?: (error: StandardError) => void): Promise<string> {
```

Comment:
Do you really want to make `merchantId` an optional param? You can if you want to for this private method. But you should move the null check from `initialize` to here, so you're throwing an error if the valid is missing. This is similar to how you handle `actions` - how it is actually a required param but you're just doing the check here instead of `initialize`. Alternatively, you might want to make both params required and do the check from the caller, but I think it might be a little awkward in this case.